File: src/main/java/core/procedures/equivalence/Eq.java
package core.procedures.equivalence;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Symbol;

import static core.scm.Cons.EMPTY;

public class Eq extends AFn {

  public Eq() {
    super(new FnArgsBuilder().min(2).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "eq?";
  }

  @Override
  public Boolean apply(Object... args) {
    Boolean result = Boolean.TRUE;
    for (int i = 0; i < args.length - 1; i++) {
      result = result && eq(args[i], args[i + 1]);
    }
    return result;
  }

  @Override
  public Boolean apply2(Object arg1, Object arg2) {
    return eq(arg1, arg2);
  }

  private boolean eq(Object first, Object second) {
    if (first instanceof Symbol && second instanceof Symbol && first != second) {
      /* Now check if 2 symbols are eq without metadata (if attached) */
      return first.equals(second);
    }
    return EMPTY.equals(first) && (EMPTY.equals(second)) || first == second;
  }
}

File: src/main/java/core/procedures/equivalence/Eq.kt
package core.procedures.equivalence

import core.procedures.AFn
import core.procedures.Arity.AtLeast
import core.scm.Symbol

class Eq : AFn<Any?, Boolean>(name = "eq?", isPure = true, arity = AtLeast(2)) {

    private val empty = emptyList<Nothing>()

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        2    -> invoke(args[0], args[1])
        else -> (0..args.size - 2).all { invoke(args[it], args[it + 1]) }
    }

    override operator fun invoke(arg1: Any?, arg2: Any?) = when {
        /* Check if 2 symbols are eq ignoring metadata */
        arg1 is Symbol && arg2 is Symbol -> arg1 == arg2
        else -> arg1 === arg2 || (empty == arg1 && empty == arg2)
    }
}
